src/coreclr/jit/importercalls.cpp

```diff
@@ -3364,6 +3366,27 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                                 retNode = gtNewFalse();
                             }
                             break;
+                        case NI_System_Type_get_IsGenericType:
+                            // a type is a generic type if there is at least one type argument that is
+                            // some valid type, so we can always just check the one at index 0 for this.
+                            // This will work on open generic types as well, and the returned type handle
+                            // will be some handle that represents the (non constructed) type parameter.
+                            if (info.compCompHnd->getTypeInstantiationArgument(hClass, 0) != NO_CLASS_HANDLE)
+                            {
+                                retNode = gtNewTrue();
+                            }
+                            else if ((info.compCompHnd->getClassAttribs(hClass) & CORINFO_FLG_SHAREDINST) != 0)
+                            {
+                                // if we have no type arguments, check that the type itself is not __Canon, and
```

Comment:
You will also get here for `__Canon[]` that it is perfectly fine to hardcode false for.

I understand that you are trying to reuse the existing JIT/EE interface APIs and flags in creative ways, but doing so often leads to bugs like https://github.com/dotnet/runtime/issues/97134#issuecomment-1897940588

Reply:
I see. I assume for the same reason we shouldn't just try to check for well known cases here, like excluding arrays and pointers types? If we should add a new JIT/EE method for this, what other VM method should it be using instead, is there one that already does what we want to do here? 🤔

Reply:
Yes, introducing JIT/EE interface method that answers the exact question tends to work the best. It can be similar to the existing `isEnum` method.